### PR TITLE
fix(core): cap average-reward array-loop sequence length before scan hang

### PR DESCRIPTION
**Pull request by rama0x1**
## Summary

`alberta_framework/core/average_reward.py`'s five `run_*_from_arrays` runners
(`run_differential_td_from_arrays`, `run_differential_gtd_from_arrays`,
`run_average_reward_horde_from_arrays`,
`run_average_reward_horde_actor_critic_from_arrays`,
`run_differential_sarsa_from_arrays`) hand their caller-supplied step arrays
(`observations`, `rewards`, `next_observations`, `cumulants`, `rhos`,
`discounts`) straight to `jax.lax.scan` with **no bound on the leading
(step) axis** and, for four of the five, no length-matching check at all. A
hostile or mistaken caller supplying a huge array forces JAX to materialize
per-step outputs (and stacked inputs) at that length, hanging the process or
exhausting memory well before any step executes.

This is the same hang class already fixed for other scan-driven array loops
in this repo (`core/sarsa.py` #1996, `core/partner_policy_fusion.py` #1991,
`utils/nexting.py` #1992, `core/learners.py` #1989) — `average_reward.py`
was not covered by any of those. All five functions are exported public API
(`alberta_framework.__init__` and `alberta_framework.core.__init__`), and
two of them (`run_differential_td_from_arrays`,
`run_differential_sarsa_from_arrays`) are the runners the Step 5/6 library
facades (`alberta_framework/steps/step5.py`, `step6.py`) call into.

## Fix

Adds `_require_avg_reward_sequence_length` (exact-type JAX-array + shape
gate, rejects a leading length outside `[1, 50_000]` before any scan) and
`_require_avg_reward_matching_length` (checks companion arrays share the
primary sequence's leading length). Applied at the top of all five
`run_*_from_arrays` functions, before `jax.lax.scan` is called.

The `50_000` ceiling is set with headroom above this module's largest
already-exercised sequence (20,000 steps, in
`test_differential_gtd_scan_learns_average_reward_cycle`) while still
bounding the previously-unbounded leading axis. `run_average_reward_horde_from_arrays`
already had per-argument shape-matching validation; this PR adds only the
missing absolute-length cap there, preserving its existing tolerant
`jnp.asarray`-based input handling.

## Scope

- `alberta_framework/core/average_reward.py`
- `tests/test_average_reward.py`

## Verification

Commit under test: `161f13a90514ae4e91bfdfdb36a822e86839437f`

```text
# On origin/main ec035f73 (pre-fix): no cap exists on any of the five
# run_*_from_arrays functions; jax.lax.scan is called unconditionally with
# the caller-supplied leading length. Confirmed by reading the pre-fix
# source (no _require_* call before jax.lax.scan in any of them).

# At this head: oversized (50_001), origin-hang-class-sized (2,000,000),
# and length-mismatched inputs are rejected with a ValueError for all five
# functions, and a monkeypatch spy proves
# alberta_framework.core.average_reward.jax.lax.scan is never invoked on a
# rejected input. Pre-existing behavior (including the 20,000-step
# in-bound GTD test) is unchanged.

.venv/bin/python -m pytest tests/test_average_reward.py -q -o addopts=""
# 114 tests passed (98 before this change, +16 new)

.venv/bin/python -m pytest tests/test_differential_sarsa_horizon.py \
  tests/test_differential_sarsa_discounts.py tests/test_step5_step6_production.py \
  tests/test_step5_hostile_validation.py -q -o addopts=""
# 226 passed -- downstream Step 5/6 callers and discount/horizon coverage unaffected

.venv/bin/python -m ruff check alberta_framework/core/average_reward.py tests/test_average_reward.py
# All checks passed!

.venv/bin/python -m mypy alberta_framework/core/average_reward.py
# Success: no issues found in 1 source file
.venv/bin/python -m mypy tests/test_average_reward.py
# 34 errors, all pre-existing on origin/main before this change (verified by
# diffing mypy output before/after; same 34 errors, none in the newly added
# test class); this PR introduces zero new mypy errors.

python -c "import alberta_framework"  # import surface stays intact
```

- [x] logs: local pytest/ruff/mypy output above; 114 collected/passed in tests/test_average_reward.py (+16 new), 226 passed in downstream Step 5/6 + discount/horizon suites, zero new mypy errors vs. origin/main baseline

- [x] domain-artifact: N/A - scan sequence-length hang guard, no IPMNIST/benchmark shard produced

- AI assistance: yes
- Model(s) used: anthropic/claude-sonnet-5
- Agent tooling: Claude Code
- Skill path: contribute-to-asi (local skill copy)
- Provenance status: self-reported

AI provider/model: Anthropic / claude-sonnet-5
Client / agent tooling: Claude Code
Attribution status: self-reported

---

### CODE REVIEW

**Comment by lalalune** [comment]
Superseded by merged PR #2060, which consolidates this resource boundary with shared scan/work budgets, combined-work checks, and the broader regression panel.
